src/FactorGraphs/Factor.php

```diff
@@ -70,44 +67,41 @@ protected function updateMessageVariable(Message $message, Variable $variable)
      */
     public function resetMarginals()
     {
-        $allValues = &$this->_messageToVariableBinding->getAllValues();
-        foreach ($allValues as &$currentVariable)
-        {
+        $allValues = $this->_messageToVariableBinding->getAllValues();
+        foreach ($allValues as $currentVariable) {
             $currentVariable->resetToPrior();
         }
     }
 
     /**
      * Sends the ith message to the marginal and returns the log-normalization constant
+     * @param $messageIndex
+     * @return
+     * @throws Exception
      */
     public function sendMessageIndex($messageIndex)
     {
         Guard::argumentIsValidIndex($messageIndex, count($this->_messages), "messageIndex");
 
-        $message = &$this->_messages[$messageIndex];
-        $variable = &$this->_messageToVariableBinding->getValue($message);
+        $message = $this->_messages[$messageIndex];
+        $variable = $this->_messageToVariableBinding->getValue($message);
         return $this->sendMessageVariable($message, $variable);
     }
 
-    protected abstract function sendMessageVariable(Message &$message, Variable &$variable);
+    protected abstract function sendMessageVariable(Message $message, Variable $variable);
 
-    public abstract function &createVariableToMessageBinding(Variable &$variable);
+    public abstract function createVariableToMessageBinding(Variable $variable);
 
-    protected function &createVariableToMessageBindingWithMessage(Variable &$variable, Message &$message)
+    protected function createVariableToMessageBindingWithMessage(Variable $variable, Message $message)
     {
-        $index = count($this->_messages);
-        $localMessages = &$this->_messages;
-        $localMessages[] = &$message;
         $this->_messageToVariableBinding->setValue($message, $variable);
-        $localVariables = &$this->_variables;
-        $localVariables[] = &$variable;
+        $this->_messages[] = $message;
+        $this->_variables[] = $variable;
         return $message;
     }
 
     public function __toString()
     {
-        return ($this->_name != null) ? $this->_name : base::__toString();
+        return ($this->_name != null) ? $this->_name : null;
```

Comment:
This is probably OK. I don't remember if there were any cases where the base class would have a non-`null` value if `_name` was `null`. Do you happen to know if any exist?